### ISSUE TITLE
Uploads not canceling

### ISSUE THREAD

**Issue by Codes4Fun** [pylonide/pylon#61]
I'm going to try and fix this myself but wanted to put this here to update status on it and in case anyone comes across this issue...

I've noticed a number of issues with canceling uploads, which turned out that the current upload is never actually canceled on the client side, rather only the UI is updated.

Easy way to reproduce is to upload a file, cancel mid upload and watch the file keep growing on the server. This causes other problems if you then try to upload again, the file becomes larger than it is supposed to be and can be corrupt.

I traced it down to some unimplemented code or possibly a bad merger. But the code in question is in this commit:

https://github.com/pylonide/pylon/commit/f4d17aec7ce7aef422173beac7ed15121a288c25#diff-c182d692a529514070174c0f5f239013

Basically that commit updated uploadfiles.js by sending a 'cancelall' message and expecting a 'canceled' message back from uploadworker.js, but in that commit uploadworker.js was not updated with support for 'cancelall' or 'canceled'.

The commit says the code came from some hotfixes branch, but I was not able to find the branch (looked at several of them), so I have no idea if this is a bad merger that didn't include uploadworker.js or incomplete code.

As a result the solution may be just to implement the missing code.

**Comment by Codes4Fun** [comment]
I fixed this issue by implementing, "cancelall", I've noticed though on the server side (plugins-server/cloud9.fs/fs/directory.js), it holds a streaming state until enough chunks have accumulated the exact total size of the file (if it goes beyond that size it gets stuck in that state). So to cancel it on the server, I send a final XmlHttpRequest and mark the size as smaller, to clear the state, otherwise if the file is uploaded again it will just keep growing the file beyond its original size.

There is a potentially other problem here, not related to canceling (doesn't affect this issue), but if a user accidentally refreshes the webpage while uploading, the server could be stuck in the streaming state as well. It does have a 10 minute timeout but a user could attempt to reupload in that period of time and get stuck with a corrupt and large file that could also be hidden. If I can reproduce that, I will create a new issue for that.
